up-streamer/src/ustreamer.rs

```diff
@@ -145,37 +190,95 @@ impl ForwardingListeners {
         out_authority: &str,
         forwarding_id: &str,
         out_sender: Sender<Arc<UMessage>>,
-    ) -> Option<Arc<ForwardingListener>> {
+        subscription_cache: Arc<Mutex<SubscriptionCache>>,
+    ) -> Result<Option<Arc<ForwardingListener>>, ForwardingListenerError> {
         let in_comparable_transport = ComparableTransport::new(in_transport.clone());
-
         let mut forwarding_listeners = self.listeners.lock().await;
 
-        let (active, forwarding_listener) = forwarding_listeners
-            .entry((in_comparable_transport.clone(), out_authority.to_string()))
-            .or_insert_with(|| {
-                let forwarding_listener = Arc::new(ForwardingListener::new(forwarding_id, out_sender));
+        if let Some((active, forwarding_listener)) = forwarding_listeners
+            .get_mut(&(in_comparable_transport.clone(), out_authority.to_string()))
+        {
+            *active += 1;
+            if *active > 1 {
+                return Ok(None);
+            } else {
+                return Ok(Some(forwarding_listener.clone()));
+            }
+        }
 
-                let reg_res = task::block_on(in_transport
-                    .register_listener(&any_uuri(), Some(&uauthority_to_uuri(out_authority)), forwarding_listener.clone()));
+        let forwarding_listener =
+            Arc::new(ForwardingListener::new(forwarding_id, out_sender.clone()));
 
-                if let Err(err) = reg_res {
-                    warn!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} unable to register listener, error: {err}");
-                } else {
-                    debug!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} able to register listener");
-                }
+        type UUriPair = Vec<(UUri, Option<UUri>)>;
```

Comment:
I think you can just define this as a tuple, since you seem to make a HashSet of them down below and only insert into the HashSet, right?

```suggestion
        type SourceSinkFilterPair = (UUri, Option<UUri>);
```

Reply:
true, done